File: sdk_samples/samples/CSharp/06_custom_stream/06_custom_stream.cs
/**
 * CARMEN Video SDK
 *
 * @category    CSharp Sample
 * @package     CARMEN Video CSharp Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * uses the custom source / image stream feature.
 * This example implements a custom source that reads images from a directory.
 */


using Carmen;
using Carmen.Anpr;
using Carmen.Video;
using CustomStream;

class Sample06_custom_stream
{
    static void EventHandlerCallback(Event e)
    {
        try
        {
            using (e)
            {
                Console.WriteLine("------------------------------------------------------");
                Console.WriteLine("Plate text: " + e.Vehicle.Plate.Text);
                Console.WriteLine("Country: " + e.Vehicle.Plate.Country);

                Console.WriteLine();
            }
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("Exception in event callback. " + ex);
        }
    }

    static void Main(string[] args)
    {
        try
        {
            if (args.Length < 2)
            {
                Console.WriteLine("Usage: <program name> <region code> <image dir path>");

                //  Region code:
                //      See in Reference Manual: Region List
                
                Console.ReadKey();
                return;
            }

            String region = args[0];
            String imageDir = args[1];

            using Anpr.AnprBuilder anprBuilder = Anpr.Builder();
            using Anpr anpr = anprBuilder
                .Type(AnprType.Local) // use LocalGo for CARMEN_GO engines
                .Build();

            using StreamProcessor.StreamProcessorBuilder streamProcessorBuilder = StreamProcessor.Builder();
            using StreamProcessor stream = streamProcessorBuilder
                .Source(new MyStreamFactory(imageDir))
                .Region(region)
                .Name("Stream 1")
                .EventCallback(EventHandlerCallback)
                .StatusChangeCallback(Utils.CommonStatusCallback) // may be omitted, the example only uses it for showing status changes
                .Anpr(anpr)
                .Build();
            
            Console.WriteLine("Please press 'q' then 'Enter' to stop stream processing.");
            
            stream.Start();
            
            while (Console.ReadLine() != "q");
            
            stream.Stop();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }

        Console.WriteLine("Please press a key to exit the program.");
        Console.ReadKey();
    }
}

File: sdk_samples/samples/C++/06_custom_stream/06_custom_stream.cpp
/**
 * CARMEN Video SDK
 *
 * @category    C++ Sample
 * @package     CARMEN Video C++ Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * uses the custom source / image stream feature.
 * This example implements a custom source that reads images from a directory.
 */

#include <filesystem>
#include <iostream>

#include <carmen/AnprBuilder.hpp>
#include <carmen/StreamProcessorBuilder.hpp>
#include <carmen/CustomImageStream.hpp>

#include "../common/utils.hpp"

#include "custom_stream.h"


void onEventCallback(const cm::Event& event) {
    std::cout << "------------------------------------------------------" << std::endl;
    std::cout << "Plate text: " << event.vehicle().plate().text() << std::endl;
    std::cout << "Country: " << event.vehicle().plate().country() << std::endl;

    std::cout << std::endl;
}


int main(int argc, char** argv) {
    try {
        if(argc < 3) {
            std::cout << "Usage: " << argv[0] << " <region code> <image dir path>" << std::endl;
            //  Region code:
            //      See in Reference Manual: Region List
            return -1;
        }

        std::string region = argv[1];
        std::string imageDir = argv[2];

        cm::anpr::Anpr anpr = cm::anpr::AnprBuilder()
                .type(cm::anpr::AnprBuilder::Type::LOCAL) // use LOCAL_GO for CARMEN_GO engines
                .build();

        cm::video::StreamProcessor stream = cm::video::StreamProcessorBuilder()
                .source(std::make_unique<MyStreamFactory>(imageDir))
                .region(region)
                .name("Stream 1")
                .eventCallback(onEventCallback)
                .anpr(anpr)
                .statusChangeCallback(commonStatusCallback) // may be omitted, the example only uses it for showing status changes
                .processingMode(cm::video::StreamProcessorMode::NonLiveStream)
                .build();

        std::cout << "Please press 'q' then 'Enter' to stop stream processing." << std::endl;

        stream.start();

        while(std::cin.get() != 'q');

        stream.stop();
    } catch(const std::exception& e) {
        std::cout << e.what() << std::endl;
        return -1;
    }
    return 0;
}
